src/Microsoft.DotNet.Darc/DarcLib/VirtualMonoRepo/CodeownersGenerator.cs

```diff
@@ -0,0 +1,149 @@
+    {
+        _vmrInfo = vmrInfo;
+        _sourceManifest = sourceManifest;
+        _localGitClient = localGitClient;
+        _fileSystem = fileSystem;
+        _logger = logger;
+    }
+
+    /// <summary>
+    /// Generates the CODEOWNERS file by gathering individual repo CODEOWNERS files.
+    /// </summary>
+    public async Task UpdateCodeowners(CancellationToken cancellationToken)
+    {
+        _logger.LogInformation("Updating {codeowners}...", VmrInfo.CodeownersPath);
+
+        var destPath = _vmrInfo.VmrPath / VmrInfo.CodeownersPath;
+
+        string header = CodeownersHeader;
+        if (_fileSystem.FileExists(destPath))
+        {
+            var content = await _fileSystem.ReadAllTextAsync(destPath);
+
+            int position = content.IndexOf(CodeownersHeader);
+            if (position != -1)
+            {
+                header = content.Substring(0, position + CodeownersHeader.Length);
+            }
+        }
+
+        _fileSystem.CreateDirectory(_fileSystem.GetDirectoryName(destPath)
+            ?? throw new Exception($"Failed to create {VmrInfo.CodeownersFileName} in {destPath}"));
+
+        using (var destStream = _fileSystem.GetFileStream(destPath, FileMode.Create, FileAccess.Write))
+        using (var writer = new StreamWriter(destStream, new UTF8Encoding(encoderShouldEmitUTF8Identifier: false)))
+        {
+            if (!string.IsNullOrEmpty(header))
+            {
+                await writer.WriteAsync(header);
+            }
+
+            foreach (ISourceComponent component in _sourceManifest.Repositories.OrderBy(m => m.Path))
+            {
+                await AddCodeownersContent(component.Path, writer, cancellationToken);
+
+                foreach (var submodule in _sourceManifest.Submodules.Where(s => s.Path.StartsWith($"{component.Path}/")))
+                {
+                    await AddCodeownersContent(submodule.Path, writer, cancellationToken);
+                }
+            }
+        }
+        
+        if (_fileSystem.GetFileInfo(destPath).Length == CodeownersHeader.Length)
+        {
+            _fileSystem.DeleteFile(destPath);
+        }
+        else
+        {
+            await _localGitClient.StageAsync(_vmrInfo.VmrPath, new string[] { VmrInfo.CodeownersPath }, cancellationToken);
+        }
+
+        _logger.LogInformation("{codeowners} updated", VmrInfo.CodeownersPath);
+    }
+
+    private async Task AddCodeownersContent(string repoPath, StreamWriter writer, CancellationToken cancellationToken)
+    {
+        // CODEOWNERS files are very restricted in size so we can safely work with them in-memory
+        var content = new List<string>();
+
+        foreach (var location in s_codeownersLocations)
+        {
+            cancellationToken.ThrowIfCancellationRequested();
+
+            var codeownersPath = _vmrInfo.VmrPath / VmrInfo.SourcesDir / repoPath / location;
+
+            if (_fileSystem.FileExists(codeownersPath))
```

Comment:
Do some repos have `CODEOWNERS` files in different locations with different content? If not we might want to break in the if to avoid duplicates

Reply:
Or if there are multiple files, we can maybe buffer them and remove duplicates before adding them to the stream

Reply:
Ah we already buffer

Reply:
If there are duplicates, there are duplicates. I wouldn't stress over this and add complexity here. I just don't think we care about that.

But reading the documentation, there's this:

> To use a `CODEOWNERS` file, create a new file called `CODEOWNERS` in the .github/, root, or docs/ directory of the repository, in the branch where you'd like to add the code owners. If `CODEOWNERS` files exist in more than one of those locations, GitHub will search for them in that order and use the first one it finds.

So I can add the `break`

Reply:
Hm, let's not, if GH changes this behaviour, we'd be fucked